### PR TITLE
Top-k accuarcy

### PR DESCRIPTION
**Pull request by brendon-boldt**
Calculate top-1 through top-k accuracy in `eval()`. I think this will be important for analysis since what want to say is that the decoding space is not amenable to beam search which would imply that the looking at top-k vs top-1 accuracy would not make much difference (simple the nearest neighbors are presumably not very good). This is much easier to implement than beam search but should hopefully provide us with an indication of how much beam search would help.

### CODE REVIEW

File: evaluation.py [hunk 1]
```diff
@@ -38,7 +38,8 @@ def eval(model, loss_function, test_iter, args, ignore_index=-100) -> Any:
     mode = model.training
     model.eval()
 
-    correct = 0
+    top_k_val = 5
```
Comment (FelixLabelle) [inline evaluation.py]: This should either be a flag or declared up top like a constant (with corresponding notation LIKE_THIS_EXAMPLE)

File: evaluation.py [hunk 2]
```diff
@@ -67,26 +68,32 @@ def batch_predict(i, x):
...
             _preds = [
                 batch_predict(i, pred_embeds)
                 for i in range(0, pred_embeds.shape[1], eval_batch_size)]
             preds = torch.cat(_preds, dim=1)
-            correct += (preds.view(-1) == batch.trg[1:,:].view(-1)).sum().item()
         else:
-            preds = scores[:-1,:,:].argmax(2).squeeze()
-            correct += (preds.view(-1) == batch.trg[1:,:].view(-1)).sum().item()
+            preds = torch.topk(scores[:-1,:,:], top_k_val, sorted=True).indices
+        correct += [
```
Comment (FelixLabelle) [inline evaluation.py]: I'm a fan of one-liners, but this is too much. It's hard to read and without playing this code in  a terminal it is hard to understand how (or if) this works properly.

Either break this down into steps in a for loop or use vectorial operations exclusively.

File: utils.py [hunk 3]
```diff
@@ -96,15 +96,18 @@ def get_nearest_neighbor(
...
+    norms = neighbor_norms.repeat(*(1,) * batch_dims, 1) * x.norm(dim=-1).unsqueeze(-1)
     dots = (
-        neighbors.unsqueeze(0).repeat(*(1,)*batch_dims, 1, 1) @ x.unsqueeze(-1)
+        neighbors.unsqueeze(0).repeat(*(1,) * batch_dims, 1, 1) @ x.unsqueeze(-1)
     ).squeeze(-1)
-    if return_indexes:
-        return (dots / norms).argmax(-1)
+    if top_k > 1:
+        topk = torch.topk(dots / norms, top_k, sorted=True)
+        return topk.indices if return_indexes else topk.values
```
Comment (FelixLabelle) [inline utils.py]: When do we need the topk values instead of the indices?

**Review by FelixLabelle — CHANGES_REQUESTED** [review]
Looks good! There are just some minor things, especially a one-liner that is a little too complicated. On top of the suggestion writing a comment above it would be appreciated

File: utils.py (hunk 3 above)
Comment (brendon-boldt) [reply utils.py]: I used them during decoding at some point, but then I switched to using the indices there too.

File: evaluation.py (hunk 2 above)
Comment (brendon-boldt) [reply evaluation.py]: So would this be preferred:
```pyhon
for k in range(1, top_k_val + 1):
  correct[k-1] = (preds[..., :k] == batch.trg[1:, :, : np.newaxis]).any(-1).sum().item()
```

Comment (FelixLabelle) [reply evaluation.py]: No, each step of the list comprehension should be assigned to a variable with a name that describes what it is with respect to the process. For example:

????? = preds[..., :k] == batch.trg[1:,:,: np,newaxis]
xxxxx = ????.any(-1)
yyyyy = xxxxx.sum().item()

Where ?????, xxxxxx, yyyy explain what the result of this step is. The code should read like your explaining the steps to me out loud

File: utils.py (hunk 3 above)
Comment (FelixLabelle) [reply utils.py]: Then could we remove that flag? We already have a log of options in the code and it's just one more thing to debug

File: evaluation.py (hunk 2 above)
Comment (brendon-boldt) [reply evaluation.py]: Too verbose?
```
for k in range(1, top_k_val + 1):
  correct_tokens = (preds[..., :k] == batch.trg[1:, :, : np.newaxis])
  correct_tokens_in_top_k = correct_tokens.any(-1)
  total_correct_tokens = correct_tokens_in_top_k.sum().item()
  correct[k-1] = total_correct_tokens
```

Comment (FelixLabelle) [reply evaluation.py]: Specifically I don't understand why you're taking the first K vectors of the predictions in an iterative fashion (keeping more and more of them)

Comment (brendon-boldt) [reply evaluation.py]: I am calculating top-k accuracy. The last dimension of `preds` holds the top `k` predictions (sorted). So if I index `preds` with `[:3]`, it will get me the model's top 3 predictions. If I then reduce with `.any(-1)`, it will tell me if any of the top 3 predictions were correct. I have to do this iteratively because, as far as I know, it is not possible to vector slice in PyTorch/NumPy (e.g., `preds[:, [:1, :2, :3]]` to get the top 1, 2, and 3).

Comment (FelixLabelle) [reply evaluation.py]: Ok, that's perfect!
